java/vector/src/main/java/org/apache/arrow/vector/NullableSmallIntVector.java

```diff
@@ -0,0 +1,326 @@
+      this(name, FieldType.nullable(Types.MinorType.SMALLINT.getType()),
+              allocator);
+   }
+
+   public NullableSmallIntVector(String name, FieldType fieldType, BufferAllocator allocator) {
+      super(name, allocator, fieldType, TYPE_WIDTH);
+      reader = new SmallIntReaderImpl(NullableSmallIntVector.this);
+   }
+
+   @Override
+   protected org.slf4j.Logger getLogger() {
+      return logger;
+   }
+
+   @Override
+   public FieldReader getReader(){
+      return reader;
+   }
+
+   @Override
+   public Types.MinorType getMinorType() {
+      return Types.MinorType.SMALLINT;
+   }
+
+
+   /******************************************************************
+    *                                                                *
+    *          vector value retrieval methods                        *
+    *                                                                *
+    ******************************************************************/
+
+
+   /**
+    * Get the element at the given index from the vector.
+    *
+    * @param index   position of element
+    * @return element at given index
+    */
+   public short get(int index) throws IllegalStateException {
+      if(isSet(index) == 0) {
+         throw new IllegalStateException("Value at index is null");
+      }
+      return valueBuffer.getShort(index * TYPE_WIDTH);
+   }
+
+   /**
+    * Get the element at the given index from the vector and
+    * sets the state in holder. If element at given index
+    * is null, holder.isSet will be zero.
+    *
+    * @param index   position of element
+    */
+   public void get(int index, NullableSmallIntHolder holder){
+      if(isSet(index) == 0) {
+         holder.isSet = 0;
+         return;
+      }
+      holder.isSet = 1;
+      holder.value = valueBuffer.getShort(index * TYPE_WIDTH);
+   }
+
+   /**
+    * Same as {@link #get(int)}.
+    *
+    * @param index   position of element
+    * @return element at given index
+    */
+   public Short getObject(int index) {
+      if (isSet(index) == 0) {
+         return null;
+      } else {
+         return get(index);
+      }
+   }
+
+   public void copyFrom(int fromIndex, int thisIndex, NullableSmallIntVector from) {
+      if (from.isSet(fromIndex) != 0) {
+         set(thisIndex, from.get(fromIndex));
+      }
+   }
+
+   public void copyFromSafe(int fromIndex, int thisIndex, NullableSmallIntVector from) {
+      handleSafe(thisIndex);
+      copyFrom(fromIndex, thisIndex, from);
+   }
+
+
+   /******************************************************************
+    *                                                                *
+    *          vector value setter methods                           *
+    *                                                                *
+    ******************************************************************/
+
+
+   private void setValue(int index, int value) {
+      valueBuffer.setShort(index * TYPE_WIDTH, value);
```

Comment:
is some kind of cast needed for value which is an int?  Why do we need to have double the set methods to allow setting an int?

Reply:
@BryanCutler 

There is probably no cast needed. I just thought that since this vector type really doesn't have 4 byte ints, we should ideally have a method set(index, short value). In the existing code on master, we don't have this. We just have set (int index, int value)

So set(index, int value) internally just takes 2 bytes from the int and calls setShort on the ArrowBuf. From an API perspective, it looked less intuitive to me since the class description tells user that the vector stores 2 byte values. So I introduced another method and still kept the original one. 

What do you think? Should we not have set(int index, short value)?

Reply:
Can we have just `setValue(int index, short value)`?

Reply:
I guess having a `setValue(int index, int value)`is fine as a convenient method . Although, do we want any overflow checks?

Reply:
I think we should just have `setValue(int index, short value)` if the other is just convenience.  Otherwise we have to think about things like overflow like @icexelloss pointed out.

Reply:
my .02: given java's penchant for upcasting, I'm inclined to include both short and int signatures. I know some of these things were also introduced because of various runtime code generation patterns that we've used. @siddharthteotia, any sense whether we rely on this pattern? (int set)

Reply:
@jacques-n , Iet me dig into the code and find out what is our usage

Reply:
Where do we stand on this?

Reply:
I think we should keep both the methods. Keeping them is probably no harm but removing them is likely going to affect the downstream use in our run time generated code.